### PR TITLE
[MRG] Fix similarity bug in NMSLIB indexer

### PR DESCRIPTION
**Pull request by piskvorky**
* Minor docstring clean up in the `similarities.nmslib` module. Follow up from https://github.com/RaRe-Technologies/gensim/issues/2820#issuecomment-623074467.
* Added `similarities.nmslib` to our API documentation – it was missing before.
* Fixed a bug where `similarities.nmslib`'s `most_similar()` returned correct neighbours but with incorrect similarities.
* Fixed deprecation warnings in Annoy: `The default argument for metric will be removed in future version of Annoy. Please pass metric='angular' explicitly.`
* Went over our API reference page https://radimrehurek.com/gensim/apiref.html (visually, as users actually see it) and fixed a bunch of misformattings and errors.

### CODE REVIEW

File: gensim/similarities/nmslib.py [hunk 1]
```diff
@@ -210,21 +219,23 @@ def _build_from_model(self, vectors, labels):
...
-            List of most similar items in format [(`item`, `cosine_distance`), ... ]
+            List of most similar items in the format `[(item, cosine_similarity), ... ].
 
         """
         ids, distances = self.index.knnQueryBatch(vector.reshape(1, -1), k=num_neighbors)[0]
 
-        return [(self.labels[ids[i]], 1 - distances[i] / 2) for i in range(len(ids))]
+        # NMSLIB returns cosine distance (not similarity), which is simply `dist = 1 - cossim`.
+        # So, convert back to similarities here.
+        return [(self.labels[ids[i]], 1 - distances[i]) for i in range(len(ids))]
```
Comment (mpenkov) [inline gensim/similarities/nmslib.py:241]: Do we need tests to cover this fix?

Comment (piskvorky) [reply gensim/similarities/nmslib.py:241]: Maybe. I'm waiting for the CI to complete, to see what tests failed :)

Let me try to find where / if nmslib is being tested.

Comment (mpenkov) [reply gensim/similarities/nmslib.py:241]: Do want to deal with this on the develop branch instead?

If this is not necessary for 3.8.3, then I'd rather release 3.8.3 without it, because it has already taken long enough. The alternative is:

1. Add tests
2. Merge onto develop
3. Merge develop into release-3.8.3 (we have done it at least once in the past)
4. Wait for CI to complete
5. Build wheels
6. Release

That will add at least another day to the release.

Comment (piskvorky) [reply gensim/similarities/nmslib.py]: No tests failed – we're testing the returned ids, but not similarities.

I think this is on purpose, because the model is trained on a tiny corpus and the exact vectors are probably unstable => similarities will differ between runs.

Do you think it's worthwhile changing the tests to be deterministic, and then testing for exact similarities too?

Comment (piskvorky) [reply gensim/similarities/nmslib.py]: > Do want to deal with this on the develop branch instead?

Like I said up to you – nothing in this PR is critical. Just another little clean up.

Comment (piskvorky) [reply gensim/similarities/nmslib.py]: > /Volumes/work/workspace/vew/tst/bin/ipython:8: DeprecationWarning: Call to deprecated `most_similar` (Method will be removed in 4.0.0, use self.wv.most_similar() instead).
  sys.exit(start_ipython())
Out[2]: [('cat', 1.000000238418579), ('woof', 0.12982195615768433)]

**Comment by piskvorky** [comment]
I cannot find the nmslib module in our API documentation at all – is this on purpose?

nmslib

That `similarities.index` showing Annoy is very weird too.

@mpenkov probably best to go ahead with 3.8.3 without this PR. Haven't reached to bottom of the rabbit hole yet.

**Comment by mpenkov** [comment]
@piskvorky The CI workers for this PR are interfering with the building of the wheels for the release. Can we kill them? Otherwise, we will have to release tomorrow, as it is the end of the day for me here.

**Comment by piskvorky** [comment]
Sure, kill what you need. We're not merging this into 3.8.3.

**Comment by piskvorky** [comment]
I had a peek at the code within `summarization` and that subpackage has to go.

**Comment by piskvorky** [comment]
@mpenkov where do I find the instructions for updating tutorials?

I looked at Wiki (https://github.com/RaRe-Technologies/gensim/wiki/How-to-author-documentation and https://github.com/RaRe-Technologies/gensim/wiki/Developer-page) but don't see it there.

I updated the `run_annoy.py` tutorial but Circle CI is failing now with timeouts. So I want to check what's wrong. Cheers.

**Comment by piskvorky** [comment]
@mpenkov ping let me know – other than the tutorials that time out, this PR is ready.

**Comment by piskvorky** [comment]
Following this comment https://github.com/RaRe-Technologies/gensim/issues/2804#issuecomment-626111610 and the docs I found [here](https://radimrehurek.com/gensim/auto_examples/howtos/run_doc.html#sphx-glr-auto-examples-howtos-run-doc-py) I tried this:

```bash
pip install -e .[docs]
make -C docs/src html
```

Some notes:
* `pip install -e .[docs]` downloads half the internet (pyemd, annoy, nmslib, nltk, sphinx, statsmodels, sklearn, cython…); seems an overkill for a user who's only interested in a single notebook.
* `make -C docs/src html` showed a ?warning? `failed: 'str' object has no attribute '__dict__'` and ultimately failed with `toctree contains reference to nonexisting document 'install'`. And also to `'license'`, `'citing'`, `'tutorial'` `'tut1'`, `'tut2'`, `'tut3'`, `'intro'` and `'design'`.
   Screen Shot 2020-05-09 at 11 16 50

    After much fixing, I realized `index.rst` is a local untracked file (some leftover from before Gallery I guess)… so I simply deleted it.

Otherwise pretty smooth, once I found the instructions and dependencies 😅 I much prefer the editing of Python files to notebooks.

**Comment by piskvorky** [comment]
@mpenkov can I leave the merge to you? This PR was originally meant for 3.8.3 so its base is ` release-3.8.3` but I guess we want to merge against `develop` now.

**Comment by piskvorky** [comment]
Travis tests fail on `TestDoc2VecModel.test_dmc_hs`, which I haven't touched in this PR:
Screen Shot 2020-05-10 at 10 29 01

@gojomo have you seen this before? Any idea what's wrong? The failures is on py2.7 only, so perhaps that test is non-deterministic? (depends on some property or RNG of python3)

**Comment by gojomo** [comment]
What I've observed is that some of these stochastic tests have shown way more variability on our cloud machines than I've ever been able to reproduce locally - even after trying to match OS/libraries as closely as possible. Toy-run parameters & thresholds that don't fail 1-in-a-500 times locally, fail closer to 1-in-2 on some of our cloud test configs. I've wondered if contributing factors slight chip/virtualization architecture issues, non-crashing memory corruption, or extreme-corner-case thread-scheduling-oddities only seen on loaded/virtualized machines. 

Typically intensifying the tests in some way has helped, but I'm not super happy with either the resulting extra cost in slower test-suite-completion, or just having the mystery of wildly-different behavior linger. At times our cloud tests have had an automatic retry-count -- I don't recall if they do right now or not -- but I'm not a fan of such retries, as it obscures useful failure feedback. (I've lost some time mistakenly thinking I'd fixed another similar issue when I didn't realize those retries were in place.)

But also: it seemed that in my *2Vec cleanup branch (#2698), some other allocation fix may have indirectly improved a bunch of these tests' stability. I've just picked that branch/PR back up, working to rebase on `develop` to ready for last steps & merge. It might offer some relief, so if a 'rerun tests' is enough to skip the error here, I'd just leave it aside for now.

Related: there's a whole bunch of cosmetic/comment/superficial changes in this PR that are unrelated to the main fix. That makes both review here harder -- have to look to make sure those `doc2vec.py` edits aren't implicated in the test failure -- & will make the rebase/merge of actual big functional changes in the same files a bit more effortful.

**Comment by piskvorky** [comment]
I did re-run and it did "fix" the CI failures. I agree it's not essential to this PR, one way or another.

> a whole bunch of cosmetic/comment/superficial changes in this PR that are unrelated to the main fix. That makes both review here harder

Not ideal, but there's no other way – at least given my energy constraints. I fix the files as I skim (most of these fixes were unplanned) and if I had to pause and open and switch between separate branches, I know I'd never do it at all.

Really looking forward to #2698 (and, by extension, 4.0.0).

File: docs/src/auto_examples/tutorials/run_annoy.py [hunk 2]
```diff
@@ -138,7 +137,7 @@ def avg_query_time(annoy_index=None, queries=1000):
         total_time += time.process_time() - start_time
     return total_time / queries
 
-queries = 10000
+queries = 1000
```
Comment (piskvorky) [inline docs/src/auto_examples/tutorials/run_annoy.py:140]: @mpenkov drawing your attention to this. Most other changes in this PR are cosmetic / documentation, but here I changed the actual value.

The idea was to speed up the tests. I saw no benefit to averaging over 10000 samples compared to 1000 samples.

**Comment by piskvorky** [comment]
@mpenkov I'm handling this one now. The PR was left for too long, the py2-vs-py3 changes are not mergeable now. I'll re-apply the relevant changes here on top of current `develop`.

**Comment by piskvorky** [comment]
Replaced by #2899.
